README.md
<h2>What is <i>JumpySaola</i></h2>  
Jumpy Saola is a <i>budget management tool</i>, which helps to track home balance by giving user possibility to track incomes and outcomes with self-declared categories and periods of time. User can add transactions and allocate them in specific limits, that may be handful to follow groups of expenses and find out which of them has the biggest impact on budget.

<h2>About Saola</h2>  
Saola <i>(Pseudoryx nghetinhensis)</i> is one of the world's rarest large mammals, found only in the Annamite Range of Vietnam and Laos. The saola is currently considered to be critically endangered. Its restrictive habitat requirements and aversion to human proximity are likely to endanger it through habitat loss and habitat fragmentation. Saola suffer losses through local hunting and the illegal trade in furs, traditional medicines, and for use of the meat in restaurants and food markets. As far as there are only few hundred specimens of this animal, hunting for their furs is lucrative local business, which leads to a gradual decline in the already small population. This animal was chosen as mascot for this project to raise awareness, that even animals, that were discovered very recently and live very remotely from humans are also endangered and people have got big impact on it.


![serval.png](doc/saola.PNG)\
<i>source:</i> [Saola - Wikipedia](https://en.wikipedia.org/wiki/Saola)


CHANGELOG
JumpySaola - release 1.2

- Added transactions and categories summary
- Added history tab
- Added chart generation for each available category
- Added chart generation for all categories in outcomes (pie view)
- Added chart generation for outcomes with limits view
- Added chart generation for outcomes with incomes
- Added chart generation for all history records
- Added possibility to run release job from specific branch
- Added possibility to deploy docker image to local docker registry
- Unit tests were cleaned up and extended
- Added basic selenium tests to cover requirements from functional tests
- Bug fixes
- Use of light weight docker image for building application was postponed for next release
- Implementation of redirect and flows tests was terminated during release process

Associated docker tag: mcieciora/jumpy_saola:1.2.102


JumpySaola - release 1.1

- Added categories for transactions
- Added periods to track transactions in specific range of time
- Added transactions' and categories' edit/remove functionalities
- Added new application layout
- Added README.md file
- Extended Jenkinsfile
- Added unit and functional tests
- Image releasing is now fully automated
- Removed index.js

Associated docker tag: mcieciora/jumpy_saola:1.1.63


JumpySaola - release 1.0

- Added Flask app initialization and database creation
- Added User and Transaction models
- Added main page views
- Added user authorization and transactions functionality
- Added Docker image build file
- Added Jenkinsfile with code linting, testing, image building, deploying and verification
- Added empty automated tests suites and initialized test structure
- Added base HTML file and home/login/signup HTMLs
- Added index.js file with transaction deletion handle

Associated docker tag: mcieciora/jumpy_saola:1.0.46

automated_tests/flows/test_flow_basic.py
from ..conftest import sign_up, log_in, log_out, add_category, edit_category, delete_category, add_transaction, \
    edit_transaction, delete_transaction, start_period, stop_period


def test_flow_basic__f_req_1(logger, client):
    """
        Covers: F-REQ1
    """
    username = "user1"
    pin_code = "1234"
    test_client = client
    sign_up(logger, test_client, dict(username=username, pin_code_1=pin_code, pin_code_2=pin_code))
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    add_category(logger, test_client, dict(category_name='category', category_limit=100))
    edit_category(logger, test_client, dict(category_name='category', category_limit=250), 1)
    delete_category(logger, test_client, 1)
    add_category(logger, test_client, dict(category_name='category', category_limit=100))
    start_period(logger, test_client, dict(period_name='new_period'))
    add_transaction(logger, test_client, dict(transaction_value='-25', transaction_desc='shopping',
                                              transaction_category='category',
                                              transaction_outcome=None))
    edit_transaction(logger, test_client, dict(transaction_value='25', transaction_desc='shopping',
                                               transaction_category='category', transaction_outcome=None), 1)
    delete_transaction(logger, test_client, 1)
    add_transaction(logger, test_client, dict(transaction_value='-25', transaction_desc='shopping',
                                              transaction_category='category',
                                              transaction_outcome=None))
    add_transaction(logger, test_client, dict(transaction_value='-50', transaction_desc='shopping',
                                              transaction_category='category',
                                              transaction_outcome='transaction_outcome'))
    stop_period(logger, test_client)
    response = client.get('/history')
    logger.info(response.data)
    assert '<td>new_period</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>-50</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>25</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<embed class="chart" type="image/svg+xml" src= data:image/svg+xml' in response.data.decode(), \
        f'Application shall inform in history view that there are periods data available\n{response.data}'


def test_flow_basic__f_req_2(logger, client_logged_in_user):
    """
        Covers: F-REQ2
    """
    username = "user1"
    pin_code = "1234"
    test_client = client_logged_in_user
    add_category(logger, test_client, dict(category_name='category', category_limit=100))
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<td>category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>100</td>' in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<td>category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>100</td>' in response.data.decode(), 'Table field is wrong or missing'
    edit_category(logger, test_client, dict(category_name='category', category_limit=150), 1)
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<td>category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>150</td>' in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<td>category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>150</td>' in response.data.decode(), 'Table field is wrong or missing'
    delete_category(logger, test_client, 1)
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<th>Category</th>' not in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>Limit</th>' not in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>Options</th>' not in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<th>Category</th>' not in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>Limit</th>' not in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>Options</th>' not in response.data.decode(), 'Table field is wrong or missing'
    add_category(logger, test_client, dict(category_name='first_category', category_limit=100))
    add_category(logger, test_client, dict(category_name='second_category', category_limit=250))
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<td>first_category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>100</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>second_category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>250</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>350</th>' in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<td>first_category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>100</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>second_category</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>250</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>350</th>' in response.data.decode(), 'Table field is wrong or missing'


def test_flow_basic__f_req_3(logger, client_with_categories):
    """
        Covers: F-REQ3
    """
    username = "user1"
    pin_code = "1234"
    test_client = client_with_categories
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">No active period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1 align="center">No active period! Go to settings and start one!</h1>' in response.data.decode(), \
        'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">No active period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    start_period(logger, test_client, dict(period_name='new_period'))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1>new_period</h1>' in response.data.decode(), 'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">Actual period: new_period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1>new_period</h1>' in response.data.decode(), 'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">Actual period: new_period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    stop_period(logger, test_client)
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1 align="center">No active period! Go to settings and start one!</h1>' in response.data.decode(), \
        'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">No active period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1 align="center">No active period! Go to settings and start one!</h1>' in response.data.decode(), \
        'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">No active period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'


def test_flow_basic__f_req_4(logger, client_with_period):
    """
        Covers: F-REQ4
    """
    username = "user1"
    pin_code = "1234"
    test_client = client_with_period
    add_transaction(logger, test_client, dict(transaction_value='-25', transaction_desc='shopping',
                                              transaction_category='category',
                                              transaction_outcome='transaction_outcome'))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<table id="transactions_table">' in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<table id="transactions_table">' in response.data.decode(), 'Table field is wrong or missing'
    edit_transaction(logger, test_client, dict(transaction_value='50', transaction_desc='transport',
                                               transaction_category='category_plus',
                                               transaction_outcome='transaction_outcome'), 1)
    response = test_client.get('/')
    logger.info(response.data)
    assert '<td>-50</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>transport</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>category_plus</td>' in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<td>-50</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>transport</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>category_plus</td>' in response.data.decode(), 'Table field is wrong or missing'
    delete_transaction(logger, test_client, 1)
    response = test_client.get('/')
    logger.info(response.data)
    assert '<table id="transactions_table">' not in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<table id="transactions_table">' not in response.data.decode(), 'Table field is wrong or missing'
    add_transaction(logger, test_client, dict(transaction_value='-25', transaction_desc='shopping',
                                              transaction_category='category',
                                              transaction_outcome='transaction_outcome'))
    add_transaction(logger, test_client, dict(transaction_value='75', transaction_desc='transport',
                                              transaction_category='category',
                                              transaction_outcome='transaction_outcome'))
    add_transaction(logger, test_client, dict(transaction_value='100', transaction_desc='income',
                                              transaction_category='category_plus',
                                              transaction_outcome=None))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<td>-25</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>-75</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>100</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>shopping</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>transport</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>income</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>0</th>' in response.data.decode(), 'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<td>-25</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>-75</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>100</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>shopping</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>transport</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>income</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<th>0</th>' in response.data.decode(), 'Table field is wrong or missing'


def test_flow_basic__f_req_5(logger, client_with_period):
    """
        Covers: F-REQ5
    """
    username = "user1"
    pin_code = "1234"
    test_client = client_with_period
    add_transaction(logger, test_client, dict(transaction_value='-25', transaction_desc='shopping',
                                              transaction_category='category',
                                              transaction_outcome='transaction_outcome'))
    add_transaction(logger, test_client, dict(transaction_value='75', transaction_desc='transport',
                                              transaction_category='category',
                                              transaction_outcome='transaction_outcome'))
    add_transaction(logger, test_client, dict(transaction_value='125', transaction_desc='income',
                                              transaction_category='category_plus',
                                              transaction_outcome=None))
    stop_period(logger, test_client)
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1 align="center">No active period! Go to settings and start one!</h1>' in response.data.decode(), \
        'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">No active period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    log_out(logger, test_client)
    log_in(logger, test_client, dict(username=username, pin_code=pin_code))
    response = test_client.get('/')
    logger.info(response.data)
    assert '<h1 align="center">No active period! Go to settings and start one!</h1>' in response.data.decode(), \
        'Table field is wrong or missing'
    response = test_client.get('/settings')
    logger.info(response.data)
    assert '<h2 align="center">No active period</h2>' in response.data.decode(), \
        'Table field is wrong or missing'
    response = test_client.get('/history')
    logger.info(response.data)
    assert '<td>125</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>-100</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<td>new_period</td>' in response.data.decode(), 'Table field is wrong or missing'
    assert '<embed class="chart" type="image/svg+xml" src= data:image/svg+xml' in response.data.decode(), \
        f'Application shall inform in history view that there are periods data available\n{response.data}'


automated_tests/units/test_auth_login.py
def test_auth_login__no_such_username(client):
    response = client.post('/login', data=dict(username='user1', pin_code='1234'), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Username does not exist.' in response.data.decode(), \
        f'Application shall not login user that does not exist\n{response.data}'


def test_auth_login__wrong_pin_code(client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code='1224'), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Incorrect pin code, try again.' in response.data.decode(), \
        f'Application shall not login user who provided wrong credentials\n{response.data}'


def test_auth_login__empty_login(client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='', pin_code='1234'), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Username does not exist.' in response.data.decode(), \
        f'Application shall not login user that does not exist\n{response.data}'


def test_auth_login__empty_pin_code(client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code=''), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Incorrect pin code, try again.' in response.data.decode(), \
        f'Application shall not login user who provided wrong credentials\n{response.data}'


def test_auth_login__empty_credentials(client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='', pin_code=''), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Username does not exist.' in response.data.decode(), \
        f'Application shall not login user that does not exist\n{response.data}'


def test_auth_login__correct_credentials(client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code='1234'), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Success!</strong> Logged in successfully!' in response.data.decode(), \
        f'Application shall login user that exists and provided correct pin code\n{response.data}'


def test_auth_login__logout(client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code='1234'), follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Success!</strong> Logged in successfully!' in response.data.decode(), \
        f'Application shall login user that exists and provided correct pin code\n{response.data}'
    response = client_users_setup.get('/logout', follow_redirects=True)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert 'Success!</strong> Logged out!' in response.data.decode(), \
        f'Application shall logout user at request\n{response.data}'


automated_tests/units/test_unit_login.py
def test_unit_login__log_in(logger, client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code='1234'), follow_redirects=True)
    logger.info(response.data)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Success!</strong> Logged in successfully!' in response.data.decode(), \
        f'Application shall log in user who exists and provided correct pin code\n{response.data}'


def test_auth_login__wrong_pin_code(logger, client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code='1224'), follow_redirects=True)
    logger.info(response.data)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Incorrect pin code, try again.' in response.data.decode(), \
        f'Application shall not login user who provided wrong credentials\n{response.data}'


def test_auth_login__empty_login(logger, client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='', pin_code='1234'), follow_redirects=True)
    logger.info(response.data)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Username does not exist.' in response.data.decode(), \
        f'Application shall not login user that does not exist\n{response.data}'


def test_auth_login__empty_pin_code(logger, client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='user1', pin_code=''), follow_redirects=True)
    logger.info(response.data)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Incorrect pin code, try again.' in response.data.decode(), \
        f'Application shall not login user who provided wrong credentials\n{response.data}'


def test_auth_login__empty_credentials(logger, client_users_setup):
    response = client_users_setup.post('/login', data=dict(username='', pin_code=''), follow_redirects=True)
    logger.info(response.data)
    assert response.status_code == 200, f'Expected response status code: 200, actual: {response.status_code}'
    assert '<strong>Warning!</strong> Username does not exist.' in response.data.decode(), \
        f'Application shall not login user that does not exist\n{response.data}'


src/website/auth.py
from flask import Blueprint, render_template, request, flash, redirect, url_for
from .models import User
from werkzeug.security import generate_password_hash, check_password_hash
from . import db
from flask_login import login_user, login_required, logout_user, current_user


auth = Blueprint('auth', __name__)


@auth.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        if current_user.is_authenticated:
            flash('You are already logged in!', category='error')
        username = request.form.get('username')
        pin_code = request.form.get('pin_code')

        user = User.query.filter_by(username=username).first()
        if user:
            if check_password_hash(user.pin_code, pin_code):
                flash('Logged in successfully!', category='success')
                login_user(user, remember=True)
                return redirect(url_for('views.home'))
            flash('Incorrect pin code, try again.', category='error')
        else:
            flash('Username does not exist.', category='error')

    return render_template("login.html", user=current_user)


@auth.route('/logout')
@login_required
def logout():
    logout_user()
    flash('Logged out!', category='success')
    return redirect(url_for('auth.login'))


@auth.route('/signup', methods=['GET', 'POST'])
def sign_up():
    if request.method == 'POST':
        if current_user.is_authenticated:
            flash('You are already logged in!', category='error')
        username = request.form.get('username')
        pin_code_1 = request.form.get('pin_code_1')
        pin_code_2 = request.form.get('pin_code_2')

        user = User.query.filter_by(username=username).first()
        if user:
            flash('User already exists.', category='error')
        elif len(username) < 3:
            flash('User name length must be greater than 3 characters.', category='error')
        elif pin_code_1 != pin_code_2:
            flash(r'Passwords do not match.', category='error')
        elif len(pin_code_1) != 4:
            flash('Password must be exactly 4 characters long.', category='error')
        else:
            new_user = User(username=username, pin_code=generate_password_hash(pin_code_1, method='sha256'))
            db.session.add(new_user)
            db.session.commit()
            login_user(new_user, remember=True)
            flash('Account created!', category='success')
            return redirect(url_for('views.home'))

    return render_template("sign_up.html", user=current_user)


src/website/models.py
from flask_login import UserMixin
from pygal import Pie, SolidGauge
from . import db


class User(db.Model, UserMixin):
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(150), unique=True)
    pin_code = db.Column(db.String(150))
    active_period = db.relationship('Period')
    transactions = db.relationship('Transaction')
    categories = db.relationship('Category')
    history = db.relationship('History')

    def get_categories(self):
        categories = {}
        for category in self.categories:
            categories[category.name] = \
                {'limit': category.limit, 'value': self.get_category_transactions_value(category.name)}
        return categories

    def get_category_transactions_value(self, category, income=False):
        total = 0
        for transaction in self.transactions:
            if transaction.category == category:
                if income and transaction.value > 0:
                    total += transaction.value
                elif income is False and transaction.value < 0:
                    total += transaction.value
        return total

    def get_total_transaction_value(self, income):
        return sum([self.get_category_transactions_value(name, income) for name, values in
                    self.get_categories().items()])

    def get_total_limit(self):
        return sum([values['limit'] for name, values in self.get_categories().items()])

    def generate_all_outcomes_by_category(self):
        gauge = Pie(inner_radius=0.5)
        gauge.title = 'All outcomes by category'
        for category_key, category_value in self.get_categories().items():
            if category_value['value'] < 0:
                gauge.add(category_key, -category_value['value'])
        return gauge.render_data_uri()

    def generate_all_outcomes_with_limits(self):
        gauge = Pie(inner_radius=0.5)
        gauge.title = 'All outcomes with general limit'
        total_limit = self.get_total_limit()
        for category_key, category_value in self.get_categories().items():
            if category_value['value'] < 0:
                gauge.add(category_key, -category_value['value'])
                total_limit += category_value['value']
        gauge.add('Remaining limit', total_limit)
        return gauge.render_data_uri()

    def generate_incomes_outcomes_plot(self):
        gauge = Pie(inner_radius=0.5)
        gauge.title = 'Incomes and outcomes'
        gauge.add('Outcomes', -self.get_total_transaction_value(False))
        gauge.add('Incomes', self.get_total_transaction_value(True))
        return gauge.render_data_uri()

    def get_month_statistics(self):
        gauge = Pie(inner_radius=0.5)
        gauge.title = 'History chart'
        for period in self.history:
            gauge.add(f'{period.name} - outcomes', -period.outcomes)
            gauge.add(f'{period.name} - incomes', period.incomes)
        return gauge.render_data_uri()

    def plot_data(self, category):
        category_data = self.get_categories()[category]
        if category_data['value'] < 0:
            category_data['value'] = category_data['value'] * -1
        gauge = SolidGauge(inner_radius=0.50)
        gauge.add(category, [{'value': category_data['value'], 'max_value': category_data['limit']}],
                  formatter=lambda x: '{:.10g} PLN'.format(x))
        return gauge.render_data_uri()


class Transaction(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    value = db.Column(db.Integer)
    description = db.Column(db.String(150))
    category = db.Column(db.String(150))
    outcome = db.Column(db.Boolean)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'))


class Category(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(150), unique=True)
    limit = db.Column(db.Integer)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'))


class Period(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(150), unique=True)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'))


class History(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(150), unique=True)
    outcomes = db.Column(db.Integer)
    incomes = db.Column(db.Integer)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'))


automated_tests/conftest.py
from os import remove
import pytest
from src.website import create_app, create_database
import logging


@pytest.fixture
def logger():
    logging.basicConfig(level=logging.INFO)
    return logging.getLogger()


@pytest.fixture
def app():
    app = create_app()
    with app.app_context():
        create_database(app)
    yield app
    remove('../src/website/db/database.db')


@pytest.fixture
def client(app):
    return app.test_client()


@pytest.fixture
def client_users_setup(client):
    for user, pin_code in {'user1': '1234', 'user2': '2233', 'userX': '0990'}.items():
        client.post('/signup', data={'username': user, 'pin_code_1': pin_code, 'pin_code_2': pin_code})

    yield client


@pytest.fixture
def client_logged_in_user(client):
    client.post('/signup', data={'username': 'user1', 'pin_code_1': '1234', 'pin_code_2': '1234'})
    client.post('/login', data={'username': 'user1', 'pin_code': '1234'})

    yield client


@pytest.fixture
def client_with_categories(client_logged_in_user):
    client_logged_in_user.post('/settings', data={'category_name': 'category', 'category_limit': '100'})
    client_logged_in_user.post('/settings', data={'category_name': 'category_plus', 'category_limit': '150'})

    yield client_logged_in_user


@pytest.fixture
def client_with_period(client_with_categories):
    client_with_categories.post('/settings', data={'period_name': 'new_period'})

    yield client_with_categories


@pytest.fixture
def client_with_transactions(client_with_period):
    client_with_period.post('/', data={'transaction_value': '25', 'transaction_desc': 'shopping',
                                       'transaction_category': 'category',
                                       'transaction_outcome': 'transaction_outcome'})

    yield client_with_period


@pytest.fixture
def client_without_history(client_with_transactions):
    client_with_transactions.post('/', data={'transaction_value': '25', 'transaction_desc': 'shopping',
                                             'transaction_category': 'category',
                                             'transaction_outcome': None})
    yield client_with_transactions


@pytest.fixture
def client_with_history(client_without_history):
    client_without_history.post('/settings', data=dict(category_name=None, category_limit=None, period_name=None),
                                follow_redirects=True)
    yield client_without_history
